include/drivers/i2c.h

```diff
@@ -337,6 +337,108 @@ struct i2c_slave_config {
 	const struct i2c_slave_callbacks *callbacks;
 };
 
+#ifdef CONFIG_I2C_STATS
+
+#include <stats/stats.h>
+
+STATS_SECT_START(i2c)
+STATS_SECT_ENTRY32(bytes_read)
+STATS_SECT_ENTRY32(bytes_written)
+STATS_SECT_ENTRY32(message_count)
+STATS_SECT_ENTRY32(transfer_call_count)
+STATS_SECT_END;
+
+STATS_NAME_START(i2c)
+STATS_NAME(i2c, bytes_read)
+STATS_NAME(i2c, bytes_written)
+STATS_NAME(i2c, message_count)
+STATS_NAME(i2c, transfer_call_count)
+STATS_NAME_END(i2c);
+
+/**
+ * @brief I2C specific device state which allows for i2c device class specific additions
+ */
+struct i2c_device_state {
+	struct device_state devstate;
+	struct stats_i2c stats;
+};
+
+static inline void i2c_xfer_stats(const struct device *dev, struct i2c_msg *msgs,
+				  uint8_t num_msgs)
+{
+	struct i2c_device_state *state =
+		CONTAINER_OF(dev->state, struct i2c_device_state, devstate);
+	uint32_t bytes_read = 0;
+	uint32_t bytes_written = 0;
+
+	STATS_INC(state->stats, transfer_call_count);
+	STATS_INCN(state->stats, message_count, num_msgs);
+	for (uint8_t _m = 0; _m < num_msgs; _m++) {
+		if (msgs[_m].flags & I2C_MSG_READ) {
+			bytes_read += msgs[_m].len;
+		}
+		if (msgs[_m].flags & I2C_MSG_WRITE) {
+			bytes_written += msgs[_m].len;
+		}
+	}
+	STATS_INCN(state->stats, bytes_read, bytes_read);
+	STATS_INCN(state->stats, bytes_written, bytes_written);
+}
+
+#define I2C_DEVICE_STATE_DEFINE(node_id, dev_name, pm_action_cb)	\
+	static struct i2c_device_state Z_DEVICE_STATE_NAME(dev_name)	\
+	__attribute__((__section__(".z_devstate"))) = {			\
+		.devstate = {							\
+			Z_DEVICE_STATE_PM_INIT(node_id, dev_name, pm_action_cb)	\
+		}								\
+	}
+
+#define I2C_INIT_FN(dev_name, init_fn)					\
+	static inline int UTIL_CAT(dev_name, _init)(const struct device *dev) \
+	{								\
+		struct i2c_device_state *state =			\
+			CONTAINER_OF(dev->state, struct i2c_device_state, devstate); \
+		stats_init(&state->stats.s_hdr, STATS_SIZE_32, 4,	\
+			   STATS_NAME_INIT_PARMS(i2c));			\
+		stats_register(dev->name, &(state->stats.s_hdr));	\
+		return init_fn(dev);					\
+	}
+
+#define I2C_DEVICE_DT_DEFINE(node_id, init_fn, pm_action_cb,		\
+			     data_ptr, cfg_ptr, level, prio,		\
+			     api_ptr, ...)				\
+	I2C_DEVICE_STATE_DEFINE(node_id, Z_DEVICE_DT_DEV_NAME(node_id), pm_action_cb); \
+	I2C_INIT_FN(Z_DEVICE_DT_DEV_NAME(node_id), init_fn)		\
+	Z_DEVICE_DEFINE(node_id, Z_DEVICE_DT_DEV_NAME(node_id),		\
+			DEVICE_DT_NAME(node_id),			\
+			&UTIL_CAT(Z_DEVICE_DT_DEV_NAME(node_id), _init), \
+			pm_action_cb,					\
+			data_ptr, cfg_ptr, level, prio,			\
+			api_ptr,					\
+			&(Z_DEVICE_STATE_NAME(Z_DEVICE_DT_DEV_NAME(node_id)).devstate), \
+			__VA_ARGS__)
+
+#else /* CONFIG_I2C_STATS */
+
+static inline void i2c_xfer_stats(const struct device *dev, struct i2c_msg *msgs,
+				  uint8_t num_msgs)
+{
+
+}
+
+#define I2C_DEVICE_DT_DEFINE(node_id, init_fn, pm_action_cb,		\
```

Comment:
I hate to bikeshed this, but since it's a rename across all drivers that will be with us "forever", I want to point out a naming issue here.

The problem I see is that a Zephyr device that implements the i2c.h API is not the same thing as a devicetree node that is an I2C device. The latter is the bus slave; the former is the bus master.

This is at least potentially confusing -- I get that there's probably not a better name, but I think this deserves good documentation to clear up when it should be used.

Reply:
bike-shedding is healthy, do not hate it.

Reply:
That's completely fair, it is here after all undocumented. I don't think its a bike shed, as I agree it could be confusing in the future. I'll see what makes sense and propose in an updated PR with doc strings

Reply:
This relates to https://github.com/zephyrproject-rtos/zephyr/pull/24885.

I expect we'll have other subsystems wanting to do something like this eventually. It's not just stats: as shown in that PR, the shell also wants to be able to enumerate all devices of a certain "type", and we haven't figured out a great way to do that yet.

Reply:
... by which I mean that, when choosing a name here, it's probably worth trying to think of a common pattern that all APIs can follow. Using `<API>_DEVICE_DT_DEFINE` as a refinement of `DEVICE_DT_DEFINE` for devices that implement `<API>` is a natural choice, so `I2C_DEVICE_DT_DEFINE` makes a lot of sense. That's why I wrote "there's probably not a better name". I think this is a good name, but we should use docstrings to avoid confusion here since you're the pioneer in this area :).

Reply:
That's great, perhaps this macro change and the device_state container can solve other issues as well, which I sort of thought it might!

Reply:
> I'll add the doc strings then, that makes the most sense.

Looks good, thanks! I'm assuming you'll add one for the INST version too? (In devicetree.h, we often will simply write `This is equivalent to DT_FOO(DT_DRV_INST(inst), ...)` when documenting `DT_INST_FOO(inst, ...)` macros; that'd be plenty for me.)